stage1/iottymux/iottymux.go

```diff
@@ -0,0 +1,599 @@
+	if err != nil {
+		log.Printf("invalid app name (%s): %v", appName, err)
+		os.Exit(254)
+	}
+
+	var r error
+	statusFile := filepath.Join(pathPrefix, appName, "endpoints")
+	switch action {
+	// attaching
+	case "auto-attach":
+		r = actionAttach(statusFile, true)
+	case "custom-attach":
+		r = actionAttach(statusFile, false)
+
+	// muxing and proxying
+	case "iomux":
+		r = actionIOMux(statusFile)
+	case "ttymux":
+		r = actionTTYMux(statusFile)
+
+	// listing
+	case "list":
+		r = actionPrint(statusFile, os.Stdout)
+
+	default:
+		r = fmt.Errorf("unknown action %q", action)
+	}
+
+	if r != nil && r != io.EOF {
+		log.FatalE("runtime failure", r)
+	}
+	os.Exit(0)
+}
+
+// actionAttach handles the attach action, either in "automatic" or "custom endpoints" mode.
+func actionAttach(statusPath string, autoMode bool) error {
+	var endpoints Targets
+	dialTimeout := 15 * time.Second
+
+	// retrieve available endpoints
+	statusFile, err := os.OpenFile(statusPath, os.O_RDONLY, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	err = json.NewDecoder(statusFile).Decode(&endpoints)
+	_ = statusFile.Close()
+	if err != nil {
+		return err
+	}
+
+	// retrieve custom attaching modes
+	customTargets := struct {
+		ttyIn  bool
+		ttyOut bool
+		stdin  bool
+		stdout bool
+		stderr bool
+	}{}
+	if !autoMode {
+		customTargets.ttyIn, _ = strconv.ParseBool(os.Getenv("STAGE2_ATTACH_TTYIN"))
+		customTargets.ttyOut, _ = strconv.ParseBool(os.Getenv("STAGE2_ATTACH_TTYOUT"))
+		customTargets.stdin, _ = strconv.ParseBool(os.Getenv("STAGE2_ATTACH_STDIN"))
+		customTargets.stdout, _ = strconv.ParseBool(os.Getenv("STAGE2_ATTACH_STDOUT"))
+		customTargets.stderr, _ = strconv.ParseBool(os.Getenv("STAGE2_ATTACH_STDERR"))
+	}
+
+	// Proxy I/O between this process and the iottymux service:
+	//  - input (stdin, tty-in) copying routines can only be canceled by process killing (ie. user detaching)
+	//  - output (stdout, stderr, tty-out) copying routines are canceled by errors when reading from remote service
+	c := make(chan error)
+	copyOut := func(w io.Writer, conn net.Conn) {
+		_, err := io.Copy(w, conn)
+		c <- err
+	}
+
+	for _, ep := range endpoints.Targets {
+		d := net.Dialer{Timeout: dialTimeout}
+		conn, err := d.Dial(ep.Domain, ep.Address)
+		if err != nil {
+			return err
+		}
+		defer conn.Close()
+		switch ep.Name {
+		case "stdin":
+			if autoMode || customTargets.stdin {
+				go io.Copy(conn, os.Stdin)
+			}
+		case "stdout":
+			if autoMode || customTargets.stdout {
+				go copyOut(os.Stdout, conn)
+			}
+		case "stderr":
+			if autoMode || customTargets.stderr {
+				go copyOut(os.Stderr, conn)
+			}
+		case "tty":
+			if autoMode || customTargets.ttyIn {
+				go io.Copy(conn, os.Stdin)
+			}
+
+			if autoMode || customTargets.ttyOut {
+				go copyOut(os.Stdout, conn)
+			} else {
+				go copyOut(ioutil.Discard, conn)
+			}
+		}
+	}
+
+	// as soon as one output copying routine fails, this unblocks and the whole process exits
+	return <-c
```

Comment:
This codepath is reached from stage0 (`rkt attach`) and all errors here mean the app is gone so further re-attach will fail at connection time. I've left a note to split this logic a bit better, as it is no immediately clear what's the context around this snippet.